sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs

```diff
@@ -57,42 +58,84 @@ protected override void Initialize(SonarAnalysisContext context)
+                    if (HasIrregularPattern(literal.Token.Text))
                     {
                         c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, literal.GetLocation()));
                     }
                 },
                 SyntaxKind.NumericLiteralExpression);
         }
 
-        private static string ClearNumberTypeSuffix(string numberLiteral)
+        internal static bool HasIrregularPattern(string numericToken)
         {
-            if (numberLiteral.EndsWith("UL", StringComparison.OrdinalIgnoreCase))
+            var splitted = StripNumericPreAndSuffix(numericToken).Split(Dot);
+
+            // ignore multiple dots.
+            if (splitted.Length > 2)
+            {
+                return false;
+            }
+
+            var groups = splitted[0].Split(Underscore).Select(s => s.Length).ToArray();
+
+            var size = NotFound;
+
+            if (groups.Length > 1)
+            {
+                size = groups[1];
+
+                // first should not be bigger.
+                if (groups[0] > size)
+                {
+                    return true;
+                }
+
+                // all should have the same size, except the first, and the size should at least be 2.
+                if (size < 2 || groups.Skip(1).Any(g => g != size))
+                {
+                    return true;
+                }
+            }
+
+            if (splitted.Length == 1)
             {
-                return numberLiteral.Substring(0, numberLiteral.Length - 2);
+                return false;
             }
 
-            if (NumericTypeSuffix.Contains(char.ToUpperInvariant(numberLiteral[numberLiteral.Length - 1])))
+            var decimals = splitted[1].Split(Underscore).Select(s => s.Length).ToArray();
+
+            if (decimals.Length == 1)
             {
-                return numberLiteral.Substring(0, numberLiteral.Length - 1);
+                return false;
             }
 
-            return numberLiteral;
+            size = size == NotFound ? decimals[0] : size;
+
+            // the last should not be bigger
+            // all should have the same size, except the last.
+            return (decimals.Last() > size)
+                || (decimals.Take(decimals.Length - 1).Any(g => g != size));
+        }
+
+        private static string StripNumericPreAndSuffix(string numericToken)
+        {
+            // hexadecimal en binary prefixes
+            var start = numericToken.StartsWith("0x", StringComparison.InvariantCultureIgnoreCase)
+                || numericToken.StartsWith("0b", StringComparison.InvariantCultureIgnoreCase) ? 2 : 0;
+
+            var length = numericToken.Length - start;
+
+            // UL suffix.
+            if (numericToken.EndsWith("UL", StringComparison.OrdinalIgnoreCase))
+            {
+                length -= 2;
+            }
+            // single suffixes
+            else if ("LDFUMldfum".IndexOf(numericToken[numericToken.Length - 1]) != NotFound)
```

Comment:
you should only strip the `F` and `D` suffix for non-binary, non-hexa numbers as they are hexa "digits"

Reply:
also, how do we treat 'e' - what's after 'e' should be like a suffix, right?

```
1.5E6m
```

Reply:
I Fully agree. Will add that too!

Reply:
Hmm, seconds thoughts:
1.234_567E2 Is obviously valid
but:
1.234_56E2
Is that wrong? should it be 0.123.456E3